### ISSUE TITLE
Sphinx.build_main no longer exists in 1.7.0

### ISSUE THREAD

**Issue by Syntaf** [sphinx-doc/sphinx#4623]
It doesn't seem there was any mention of this in the changelog, and i'm not aware of a replacement for `sphinx.build_main`. What can I replace this command with for > 1.7.0?

### Problem
- As of 1.7.0, `sphinx.build_main` no longer exists and present a problem for repositories which use this function to build documentation within a script.

#### Procedure to reproduce the problem
```
import sphinx
sphinx.build_main(args)
```

#### Error logs / results
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'sphinx' has no attribute 'build_main'

```

#### Expected results

Pre 1.7, this function was used inside https://github.com/Syntaf/travis-sphinx to build docs and auto deploy to github web pages.

### Reproducible project / your project
- https://github.com/Syntaf/travis-sphinx

### Environment info
- Python version: 3
- Sphinx version: 1.7.0

**Comment by tk0miya** [comment]
Oh, sorry. The removal is not intended. Certainly, structures has been changed since 1.7. But, according to our deprecation policy, we should not have removed it immediately.
I'd like to add it again in next release. But it will emit deprecation warning.
So I propose you to switch to `sphinx.cmd.build.buid_main()`

**Comment by kohr-h** [comment]
There seem to be more changes in the `build_main` function. Before 1.7, calling `build_main(['-b html', srcdir, outdir])` would work, but it fails in 1.7 with the error
```
Builder name  html not registered or available through entry point
```
Note the extra space before "html", I think that's why it's not recognized as "html" builder. Probably the parser just strips off the `'-b'` and forgets to strip away whitespace.

A fix is to instead invoke `build_main(['-b', 'html', srcdir, outdir])`.

**Comment by tk0miya** [comment]
The argv format is used for the argument of `build_main()`. I think `'-b html'` is not an argv format. It is same as `sphinx-build '-b html' srcdir outdir`.
Certainly, we switched the option parser from optparse to argparse. The behavior might be changed at that time. But I don't think it is a bug of `build_main()`.

**Comment by kohr-h** [comment]
> Certainly, we switched the option parser from optparse to argparse. The behavior might be changed at that time. But I don't think it is a bug of build_main().

Yes, I was curious what kind of change that was. I didn't assume it was a bug, just a change of behavior that requires the options to be passed slightly differently.

Note that I did test `build_main(['-b', 'html', srcdir, outdir])` with Sphinx 1.6, and that didn't work. So there one had to use `'-b html'` (or something else, at least not `'-b', 'html'`).

**Comment by tk0miya** [comment]
>Note that I did test build_main(['-b', 'html', srcdir, outdir]) with Sphinx 1.6, and that didn't work. So there one had to use '-b html' (or something else, at least not '-b', 'html').

Until Sphinx-1.6, `build_main()` ignores the first argument (see #4615). so `build_main(['-b', 'html', srcdir, outdir])` is same as `sphinx-build html srcdir outdir`.
On the other hand, `build_main(['-b html', srcdir, outdir])` is same as `sphinx-build srcdir outdir`.

Note: new interface since 1.7 (`sphinx.cmd.build.buid_main()`) does not ignores the first argument.

**Comment by kohr-h** [comment]
> Until Sphinx-1.6, build_main() ignores the first argument (see #4615). so build_main(['-b', 'html', srcdir, outdir]) is same as sphinx-build html srcdir outdir.
> On the other hand, build_main(['-b html', srcdir, outdir]) is same as sphinx-build srcdir outdir.

Oh, that explains it. Thanks!



> Note: new interface since 1.7 (sphinx.cmd.build.buid_main()) does not ignores the first argument.

Alright, then `travis-sphinx` should only provide the extra option for sphinx 1.7 and higher.

**Comment by tk0miya** [comment]
I'll fix both problem in #4624. But `sphinx.main()` is still marked as deprecated. so I recommend you to move `sphinx.cmd.build.build_main()`.

Note: new `build_main()` is also not a public interface. so we can't promise you that it's stable.

**Comment by tk0miya** [comment]
Fixed #4624. Thank you for reporting!
